chapters/technology-research/state-management.tex

```diff
@@ -0,0 +1,352 @@
+    \label{code:state-mobx}
+    \begin{minted}{dart}
+    import 'package:mobx/mobx.dart';
+
+    part 'counter.g.dart';
+    
+    class Counter = CounterBase with _$Counter;
+    
+    abstract class CounterBase with Store {
+        @observable
+        int value = 0;
+    
+        @action
+        void increment() {
+            value++;
+        }
+    }
+    \end{minted}
+\end{listing}
+
+Z výpisu kódu \ref{code:state-mobx} lze vypozorovat,
+že knihovna pracuje s generovanými kódy z pomocných generátorů.
+Tím se dle \cite{mobx_core_concepts} snaží redukovat množství kódu,
+které by bylo jinak vyžadováno napsat k docílení stejné funkčnosti.
+Knihovna také využívá anotace,
+kterými je kód zpřehledněn.
+Generování kódu však není nejrychlejší,
+což je jedna z nevýhod tohoto přístupu.
+
+\subsection{Redux}
+
+\todo{Napsat nějaký větší úvod o knihovně Redux?}
+Jedna z nejznámějších knihoven řešící state management je knihovna Redux.
+Dle \cite{redux_basics} pracuje Redux se třemi základními koncepty:
+
+\begin{description}
+    \item[Actions] Obsahují data,
+    která se posílají z aplikace do \emph{Store},
+    a jsou jediným zdrojem informací.
+    \item[Reducers] Popisují jak se změní stav aplikace v závislosti na akci.
+    \item[Store] Pojí předešlé dva koncepty dohromady.
+    Drží stav aplikace,
+    určuje počáteční stav
+    a poskytuje aktuální stav.
+\end{description}
+
+\begin{listing}
+    \caption{Ukázka kódu počítadla v knihovně Redux \cite{redux_basics}}
+    \label{code:state-redux}
+    \begin{minted}{dart}
+    import 'package:redux/redux.dart';
+
+    enum CounterActions {
+        increment
+    }
+    
+    class AppState {
+        int value;
+    
+        AppState({ 
+            this.value = 0,
+        });
+    }
+    
+    AppState counterReducer(AppState state, action) {
+        switch (action) {
+            case CounterActions.increment:
+                return AppState(
+                    value: state.value++,
+                );
+            default:
+                return state;
+        }
+    }
+    \end{minted}
+\end{listing}
+
+Dle \cite{redux_basics} knihovna Redux obsahuje právě jeden \emph{Store}
+a členění se provádí pomocí kompozice \emph{Reducers}.
+Z výpisu kódu \ref{code:state-redux} je vidět,
+že ačkoli je zápis přehledný,
+ve větší aplikaci nemusí být omezení pouze jednoho \emph{Store} vhodné.
+
+\subsection{BLoC}
+
+Oproti předchozím příkladům není Business Logic Component (dále jen BLoC)
+knihovna,
+ale návrhový vzor.
+Tento vzor je tvořen několika jednoduchými pravidly,
+kterými se každá implementace musí řídit.
+Přičemž konkrétní prvky implementace jsou přenechány na vývojáře.
+Pojmem BLoC také označujeme třídu,
+která uchovává stav dodržuje daný vzor. 
+Dle \cite{googledevelopers_bloc} těmito pravidly pro návrh jsou:
+\todo{Tato pravidla překládám (sám) z toho videa,
+mám to nějak dál rozepsat,
+nebo tato citace stačí?}
```

Comment:
Na toto jsem se ptal v rámci DPR a prý to mám takto vyřešeno nejlépe jak to jde, takže takto ponechám.